File: src/main/java/dark/leech/text/action/export/Text.java
package dark.leech.text.action.export;

import java.util.List;

import dark.leech.text.action.Log;
import dark.leech.text.listeners.ProgressListener;
import dark.leech.text.models.Chapter;
import dark.leech.text.models.Properties;
import dark.leech.text.models.Trash;
import dark.leech.text.util.FileUtils;
import dark.leech.text.util.HtmlSanitizer;
import dark.leech.text.util.RegexUtils;
import dark.leech.text.util.SettingUtils;
import dark.leech.text.util.SyntaxUtils;
import dark.leech.text.util.TypeUtils;

/**
 * Exports ebook content to text or HTML format. Supports both split (separate files) and combined
 * (single file) output modes with optional table of contents generation.
 *
 * @author Long
 * @since 9/17/2016
 */
public class Text {
    private static final String RAW_FILE_EXTENSION = ".txt";
    private static final String INTRODUCTION_ID = "gioithieu";
    private static final String INTRODUCTION_NAME = "Giới Thiệu";
    private static final String TABLE_OF_CONTENTS_TITLE = "Mục lục";
    private static final String PROGRESS_MESSAGE_FORMAT = "[2/3]Xuất text...";
    private static final int PROGRESS_MULTIPLIER = 100;

    private final Properties properties;
    private final ExportFormat format;
    private final ExportMode exportMode;
    private final ExportOptions options;
    private final String charset;

    private ProgressListener progressListener;
    private String templateSyntax;

    public Text(Properties properties, int type, boolean makeToc, boolean includeCss, int tach) {
        this.properties = properties;
        this.format = (type == TypeUtils.HTML) ? ExportFormat.HTML : ExportFormat.TEXT;
        this.exportMode = (tach == 1) ? ExportMode.COMBINED : ExportMode.SPLIT;
        this.options = new ExportOptions(makeToc, includeCss);
        this.charset = properties.getCharset();
    }

    public void addProgressListener(ProgressListener progressListener) {
        this.progressListener = progressListener;
    }

    /** Main entry point to export content. */
    public void export() {
        initializeTemplate();
        String fileExtension = format.getFileExtension();
        exportContent(fileExtension);
    }

    private void initializeTemplate() {
        if (format == ExportFormat.HTML) {
            templateSyntax = SettingUtils.HTML_SYNTAX.replaceAll("[Uu][Tt][Ff]-8", charset);
        } else {
            templateSyntax = SettingUtils.TXT_SYNTAX;
        }
    }

    private void exportContent(String fileExtension) {
        List<Chapter> chapters = properties.getChapList();
        StringBuilder combinedContent = new StringBuilder();

        if (exportMode == ExportMode.COMBINED) {
            combinedContent.append(createCombinedFileHeader());
        }

        if (properties.isAddGt()) {
            exportIntroduction(chapters, fileExtension, combinedContent);
        }

        if (options.makeToc && format == ExportFormat.HTML) {
            exportTableOfContents(chapters, fileExtension, combinedContent);
        }

        exportChapters(chapters, fileExtension, combinedContent);

        if (exportMode == ExportMode.COMBINED) {
            finalizeCombinedFile(combinedContent, fileExtension);
        }
    }

    private String createCombinedFileHeader() {
        if (format != ExportFormat.HTML) {
            return "";
        }

        String header = templateSyntax;
        header =
                header.replaceAll(
                        "<title>.*?</title>", "<title>" + properties.getName() + "</title>");
        header = header.replaceAll("(?s)(.*?<body.*?>).*", "$1");

        if (options.includeCss) {
            header =
                    header.replace(
                            "</head>",
                            "<style>\n" + SettingUtils.CSS_SYNTAX + "\n</style>\n</head>");
        }

        return header;
    }

    private void exportIntroduction(
            List<Chapter> chapters, String fileExtension, StringBuilder combinedContent) {
        Chapter introduction = createIntroductionChapter();

        try {
            String content = processChapterContent(introduction, false);
            saveChapterContent(content, introduction.getId(), fileExtension, combinedContent);
        } catch (Exception e) {
            Log.add("Failed to export introduction: " + e.getMessage());
        }
    }

    private Chapter createIntroductionChapter() {
        Chapter chapter = new Chapter();
        chapter.setChapName(INTRODUCTION_NAME);
        chapter.setId(INTRODUCTION_ID);
        return chapter;
    }

    private void exportTableOfContents(
            List<Chapter> chapters, String fileExtension, StringBuilder combinedContent) {
        String tableOfContents = buildTableOfContentsHtml(chapters);

        if (exportMode == ExportMode.SPLIT) {
            String tocPath = properties.getSavePath() + "/data/Text/mucluc.html";
            FileUtils.string2file(tableOfContents, tocPath, charset);
        } else {
            combinedContent.append(tableOfContents);
        }
    }

    private String buildTableOfContentsHtml(List<Chapter> chapters) {
        StringBuilder tocBuilder = new StringBuilder();
        tocBuilder.append("\n<h4>" + TABLE_OF_CONTENTS_TITLE + "</h4>\n");

        if (exportMode == ExportMode.SPLIT) {
            tocBuilder.append(createSplitTableOfContents(chapters));
            return wrapInHtmlDocument(tocBuilder.toString());
        } else {
            tocBuilder.append(createCombinedTableOfContents(chapters));
            return tocBuilder.toString();
        }
    }

    private String createSplitTableOfContents(List<Chapter> chapters) {
        StringBuilder toc = new StringBuilder();

        if (properties.isAddGt()) {
            toc.append("<div class=\"lv2\"><a href=\"../Text/")
                    .append(INTRODUCTION_ID)
                    .append(".html\">")
                    .append(INTRODUCTION_NAME)
                    .append("</a></div>\n");
        }

        for (Chapter chapter : chapters) {
            toc.append("<div class=\"lv2\"><a href=\"../Text/")
                    .append(chapter.getId())
                    .append(".html\">")
                    .append(formatChapterLinkText(chapter))
                    .append("</a></div>\n");
        }

        return toc.toString();
    }

    private String createCombinedTableOfContents(List<Chapter> chapters) {
        StringBuilder toc = new StringBuilder();

        if (properties.isAddGt()) {
            toc.append("<div class=\"lv2\"><a href=\"#")
                    .append(INTRODUCTION_ID)
                    .append("\">")
                    .append(INTRODUCTION_NAME)
                    .append("</a></div>\n");
        }

        for (Chapter chapter : chapters) {
            toc.append("<div class=\"lv2\"><a href=\"#")
                    .append(chapter.getId())
                    .append("\">")
                    .append(formatChapterLinkText(chapter))
                    .append("</a></div>\n");
        }

        return toc.toString();
    }

    private String formatChapterLinkText(Chapter chapter) {
        String partName = chapter.getPartName();
        String chapterName = chapter.getChapName();

        if (partName == null || partName.isEmpty()) {
            return chapterName;
        }

        return partName + " - " + chapterName;
    }

    private String wrapInHtmlDocument(String bodyContent) {
        String document = templateSyntax;
        document =
                document.replaceAll(
                        "<title>.*?</title>", "<title>" + properties.getName() + "</title>");
        document = document.replaceAll("(?s)(.*?<body.*?>).*", "$1");
        return document + bodyContent + "</body>\n</html>";
    }

    private void exportChapters(
            List<Chapter> chapters, String fileExtension, StringBuilder combinedContent) {
        int completedCount = 0;
        int totalChapters = properties.getSize();

        for (Chapter chapter : chapters) {
            if (!chapter.isCompleted()) {
                continue;
            }

            try {
                String content = processChapterContent(chapter, true);
                saveChapterContent(content, chapter.getId(), fileExtension, combinedContent);
                completedCount++;

                reportProgress(completedCount, totalChapters);
            } catch (Exception e) {
                Log.add(
                        "Failed to export chapter "
                                + chapter.getChapName()
                                + ": "
                                + e.getMessage());
            }
        }
    }

    private void reportProgress(int completed, int total) {
        if (progressListener != null) {
            int percentage = completed * PROGRESS_MULTIPLIER / total;
            progressListener.setProgress(percentage, PROGRESS_MESSAGE_FORMAT);
        }
    }

    private void saveChapterContent(
            String content, String chapterId, String fileExtension, StringBuilder combinedContent) {

        if (exportMode == ExportMode.SPLIT) {
            String filePath = properties.getSavePath() + "/data/Text/" + chapterId + fileExtension;
            FileUtils.string2file(content, filePath, charset);
        } else {
            String processedContent = extractBodyContent(content);
            if (format == ExportFormat.TEXT) {
                processedContent += "\n\n";
            }
            combinedContent.append(processedContent);
        }
    }

    private String extractBodyContent(String htmlContent) {
        if (format == ExportFormat.HTML) {
            return htmlContent.replaceAll("(?s).*?<body.*?>(.*?)</body>.*", "$1");
        }
        return htmlContent;
    }

    private void finalizeCombinedFile(StringBuilder combinedContent, String fileExtension) {
        if (format == ExportFormat.HTML) {
            combinedContent.append("</body>\n</html>");
        }

        String outputPath = properties.getSavePath() + "/out/text" + fileExtension;
        FileUtils.string2file(combinedContent.toString(), outputPath, charset);
    }

    /**
     * Processes chapter content by loading raw text, applying transformations, and formatting.
     *
     * @param chapter Chapter to process
     * @param preserveParagraphs Whether to apply paragraph formatting
     * @return Formatted chapter content
     * @throws Exception if chapter file cannot be read
     */
    private String processChapterContent(Chapter chapter, boolean preserveParagraphs)
            throws Exception {
        String rawContent = loadRawChapterContent(chapter.getId());
        String cleanedContent = cleanTextContent(rawContent);
        String formattedContent = formatChapterContent(chapter, cleanedContent, preserveParagraphs);

        if (format == ExportFormat.HTML) {
            formattedContent = HtmlSanitizer.sanitize(formattedContent);
        }

        return formattedContent;
    }

    private String loadRawChapterContent(String chapterId) throws Exception {
        String rawFilePath = properties.getSavePath() + "/raw/" + chapterId + RAW_FILE_EXTENSION;
        return FileUtils.file2string(rawFilePath, charset);
    }

    private String cleanTextContent(String content) {
        for (Trash trashRule : SettingUtils.TRASH) {
            if (trashRule.isReplace()) {
                String searchPattern = unescapeTrashPattern(trashRule.getSrc());
                String replacement = unescapeTrashPattern(trashRule.getTo());
                content = content.replaceAll(searchPattern, replacement);
            }
        }
        return SyntaxUtils.covertString(content);
    }

    private String unescapeTrashPattern(String pattern) {
        if (pattern == null) {
            return "";
        }
        return pattern.replace("\\n", "\n").replace("\\r", "\r").replace("\\t", "\t");
    }

    private String formatChapterContent(
            Chapter chapter, String content, boolean preserveParagraphs) {

        String template = templateSyntax;

        // Replace placeholders with actual values
        template = replaceTemplatePlaceholder(template, properties.getName(), "\\[1\\]", "NAME");
        template =
                replaceTemplatePlaceholder(template, properties.getAuthor(), "\\[2\\]", "AUTHOR");
        template =
                replaceTemplatePlaceholder(template, chapter.getPartName(), "\\[3\\]", "NAME_PART");
        template =
                replaceTemplatePlaceholder(template, chapter.getChapName(), "\\[4\\]", "NAME_CHAP");
        template = template.replace("[ID]", chapter.getId());

        // Format paragraphs
        String formattedContent = formatParagraphs(content, preserveParagraphs);
        String paragraphPlaceholder =
                RegexUtils.find(template, "(\\[5\\].*\\[PARAGRAPH\\].*\\[5\\])", 1);

        return template.replace(paragraphPlaceholder, formattedContent);
    }

    private String replaceTemplatePlaceholder(
            String template, String value, String placeholderTag, String placeholderKey) {

        if (value == null || value.length() < 2) {
            return template.replaceAll("\\s*" + placeholderTag + ".*?" + placeholderTag, "");
        }

        String pattern =
                placeholderTag + "(.*?" + "\\[" + placeholderKey + "\\]" + ".*?)" + placeholderTag;
        String replacement = template.replaceAll(pattern, "$1");
        return replacement.replace("[" + placeholderKey + "]", value);
    }

    private String formatParagraphs(String content, boolean preserveParagraphs) {
        if (content.isEmpty()) {
            return content;
        }

        content = applyDropCapFormatting(content);

        if (preserveParagraphs) {
            String paragraphTemplate = extractParagraphTemplate();
            content = wrapInParagraphTags(content, paragraphTemplate);
        }

        return content;
    }

    private String extractParagraphTemplate() {
        String firstTag =
                RegexUtils.find(templateSyntax, "\\[5\\](.*)\\[PARAGRAPH\\](.*)\\[5\\]", 1);
        String lastTag =
                RegexUtils.find(templateSyntax, "\\[5\\](.*)\\[PARAGRAPH\\](.*)\\[5\\]", 2);
        return firstTag + "\n" + lastTag;
    }

    private String wrapInParagraphTags(String content, String paragraphTemplate) {
        String[] parts = paragraphTemplate.split("\n");
        if (parts.length != 2) {
            return content;
        }

        String firstTag = parts[0];
        String lastTag = parts[1];

        content = content.replace("\n", lastTag + "\n" + firstTag);
        return firstTag + content + lastTag;
    }

    private String applyDropCapFormatting(String content) {
        if (content.isEmpty()) {
            return content;
        }

        if (!SettingUtils.IS_DROP_SELECTED || format != ExportFormat.HTML) {
            return content;
        }

        char firstChar = content.charAt(0);
        if (firstChar == Character.toLowerCase(firstChar)) {
            return content;
        }

        String dropCapSyntax = SettingUtils.DROP_SYNTAX;
        return dropCapSyntax.replace("[DROP]", String.valueOf(firstChar)) + content.substring(1);
    }

    /** Export format enumeration. */
    private enum ExportFormat {
        HTML(".html"),
        TEXT(".txt");

        private final String fileExtension;

        ExportFormat(String fileExtension) {
            this.fileExtension = fileExtension;
        }

        public String getFileExtension() {
            return fileExtension;
        }
    }

    /** Export mode enumeration. */
    private enum ExportMode {
        SPLIT, // Separate files for each chapter
        COMBINED // Single combined file
    }

    /** Export options configuration. */
    private static class ExportOptions {
        final boolean makeToc;
        final boolean includeCss;

        ExportOptions(boolean makeToc, boolean includeCss) {
            this.makeToc = makeToc;
            this.includeCss = includeCss;
        }
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/action/export/Text.kt
package dev.haipham22.leechtext.action.export

import dev.haipham22.leechtext.log.EngineLogger
import dev.haipham22.leechtext.models.Chapter
import dev.haipham22.leechtext.models.Properties
import dev.haipham22.leechtext.models.Trash
import dev.haipham22.leechtext.util.AppSettings
import dev.haipham22.leechtext.util.TypeUtils
import dev.haipham22.leechtext.util.normalizeVietnamese
import dev.haipham22.leechtext.util.readTextOrNull
import dev.haipham22.leechtext.util.regexFind
import dev.haipham22.leechtext.util.sanitizeHtml
import dev.haipham22.leechtext.util.stripParagraphTags
import dev.haipham22.leechtext.util.writeTo
import okio.IOException
import okio.Path.Companion.toPath

/** Hậu tố link chương trong TOC HTML (format bản gốc). */
private const val CHAPTER_LINK_SUFFIX = "</a></div>\n"

/**
 * Xuất nội dung sang text/HTML (port từ action/export/Text.java).
 * Split (mỗi chương 1 file) hoặc combined (1 file duy nhất).
 */
@Suppress("LongParameterList") // tham số log DI + params nghiệp vụ — bundle làm code khó đọc
class Text(
    private val log: EngineLogger,
    private val properties: Properties,
    type: Int,
    private val makeToc: Boolean,
    private val includeCss: Boolean,
    tach: Int,
    private val settings: AppSettings,
    private val progressListener: ProgressListener? = null,
) {
    private enum class ExportFormat(
        val fileExtension: String,
    ) {
        HTML(".html"),
        TEXT(".txt"),
    }

    private enum class ExportMode { SPLIT, COMBINED }

    private val format: ExportFormat = if (type == TypeUtils.HTML) ExportFormat.HTML else ExportFormat.TEXT
    private val exportMode: ExportMode = if (tach == 1) ExportMode.COMBINED else ExportMode.SPLIT
    private val charset: String = properties.charset ?: "UTF-8"
    private var templateSyntax: String = ""

    /** Entry point. */
    fun export() {
        initializeTemplate()
        exportContent(format.fileExtension)
    }

    private fun initializeTemplate() {
        templateSyntax =
            if (format == ExportFormat.HTML) {
                settings.htmlSyntax.replace("[Uu][Tt][Ff]-8".toRegex(), charset)
            } else {
                settings.txtSyntax
            }
    }

    private fun exportContent(fileExtension: String) {
        val chapters = properties.chapList ?: emptyList()
        val combinedContent = StringBuilder()

        if (exportMode == ExportMode.COMBINED) {
            combinedContent.append(createCombinedFileHeader())
        }

        if (properties.addGt) {
            exportIntroduction(fileExtension, combinedContent)
        }

        if (makeToc && format == ExportFormat.HTML) {
            exportTableOfContents(chapters, combinedContent)
        }

        exportChapters(chapters, fileExtension, combinedContent)

        if (exportMode == ExportMode.COMBINED) {
            finalizeCombinedFile(combinedContent, fileExtension)
        }
    }

    private fun createCombinedFileHeader(): String {
        if (format != ExportFormat.HTML) return ""

        var header = templateSyntax
        header = header.replace("<title>.*?</title>".toRegex(), "<title>" + properties.name + "</title>")
        header = header.replace("(?s)(.*?<body.*?>).*".toRegex(), "$1")

        if (includeCss) {
            header =
                header.replace(
                    "</head>",
                    "<style>\n" + settings.cssSyntax + "\n</style>\n</head>",
                )
        }
        return header
    }

    private fun exportIntroduction(
        fileExtension: String,
        combinedContent: StringBuilder,
    ) {
        val introduction =
            Chapter().apply {
                chapName = INTRODUCTION_NAME
                id = INTRODUCTION_ID
            }
        try {
            val content = processChapterContent(introduction, preserveParagraphs = false)
            saveChapterContent(content, introduction.id ?: "", fileExtension, combinedContent)
        } catch (e: Exception) {
            log.add("Failed to export introduction: " + e.message)
        }
    }

    private fun exportTableOfContents(
        chapters: List<Chapter>,
        combinedContent: StringBuilder,
    ) {
        val tableOfContents = buildTableOfContentsHtml(chapters)
        if (exportMode == ExportMode.SPLIT) {
            tableOfContents.writeTo(properties.savePath + "/data/Text/mucluc.html", charset, log)
        } else {
            combinedContent.append(tableOfContents)
        }
    }

    private fun buildTableOfContentsHtml(chapters: List<Chapter>): String {
        val tocBuilder = StringBuilder()
        tocBuilder.append("\n<h4>" + TABLE_OF_CONTENTS_TITLE + "</h4>\n")
        return if (exportMode == ExportMode.SPLIT) {
            tocBuilder.append(createSplitTableOfContents(chapters))
            wrapInHtmlDocument(tocBuilder.toString())
        } else {
            tocBuilder.append(createCombinedTableOfContents(chapters))
            tocBuilder.toString()
        }
    }

    private fun createSplitTableOfContents(chapters: List<Chapter>): String = buildString {
        if (properties.addGt) {
            append("<div class=\"lv2\"><a href=\"../Text/")
                .append(INTRODUCTION_ID)
                .append(".html\">")
                .append(INTRODUCTION_NAME)
                .append(CHAPTER_LINK_SUFFIX)
        }
        for (chapter in chapters) {
            append("<div class=\"lv2\"><a href=\"../Text/")
                .append(chapter.id)
                .append(".html\">")
                .append(formatChapterLinkText(chapter))
                .append(CHAPTER_LINK_SUFFIX)
        }
    }

    private fun createCombinedTableOfContents(chapters: List<Chapter>): String = buildString {
        if (properties.addGt) {
            append("<div class=\"lv2\"><a href=\"#")
                .append(INTRODUCTION_ID)
                .append("\">")
                .append(INTRODUCTION_NAME)
                .append(CHAPTER_LINK_SUFFIX)
        }
        for (chapter in chapters) {
            append("<div class=\"lv2\"><a href=\"#")
                .append(chapter.id)
                .append("\">")
                .append(formatChapterLinkText(chapter))
                .append(CHAPTER_LINK_SUFFIX)
        }
    }

    private fun formatChapterLinkText(chapter: Chapter): String {
        val partName = chapter.partName ?: ""
        val chapterName = chapter.chapName ?: ""
        return if (partName.isEmpty()) chapterName else "$partName - $chapterName"
    }

    private fun wrapInHtmlDocument(bodyContent: String): String {
        var document = templateSyntax
        document = document.replace("<title>.*?</title>".toRegex(), "<title>" + properties.name + "</title>")
        document = document.replace("(?s)(.*?<body.*?>).*".toRegex(), "$1")
        return document + bodyContent + "</body>\n</html>"
    }

    private fun exportChapters(
        chapters: List<Chapter>,
        fileExtension: String,
        combinedContent: StringBuilder,
    ) {
        var completedCount = 0
        val totalChapters = properties.size

        for (chapter in chapters) {
            if (!chapter.completed) continue
            try {
                val content = processChapterContent(chapter, preserveParagraphs = true)
                saveChapterContent(content, chapter.id ?: "", fileExtension, combinedContent)
                completedCount++
                reportProgress(completedCount, totalChapters)
            } catch (e: Exception) {
                log.add("Failed to export chapter " + chapter.chapName + ": " + e.message)
            }
        }
    }

    private fun reportProgress(
        completed: Int,
        total: Int,
    ) {
        if (progressListener != null && total > 0) {
            progressListener.setProgress(completed * PROGRESS_MULTIPLIER / total, PROGRESS_MESSAGE_FORMAT)
        }
    }

    private fun saveChapterContent(
        content: String,
        chapterId: String,
        fileExtension: String,
        combinedContent: StringBuilder,
    ) {
        if (exportMode == ExportMode.SPLIT) {
            content.writeTo(properties.savePath + "/data/Text/" + chapterId + fileExtension, charset, log)
        } else {
            var processedContent = extractBodyContent(content)
            if (format == ExportFormat.TEXT) processedContent += "\n\n"
            combinedContent.append(processedContent)
        }
    }

    private fun extractBodyContent(htmlContent: String): String = if (format == ExportFormat.HTML) {
        htmlContent.replace("(?s).*?<body.*?>(.*?)</body>.*".toRegex(), "$1")
    } else {
        htmlContent
    }

    private fun finalizeCombinedFile(
        combinedContent: StringBuilder,
        fileExtension: String,
    ) {
        if (format == ExportFormat.HTML) {
            combinedContent.append("</body>\n</html>")
        }
        combinedContent.toString().writeTo(properties.savePath + "/out/text" + fileExtension, charset, log)
    }

    /**
     * Load raw text → áp regex Trash + normalizeVietnamese → format theo template.
     *
     * @param preserveParagraphs có wrap `<p>` theo template [5] hay không
     */
    private fun processChapterContent(
        chapter: Chapter,
        preserveParagraphs: Boolean,
    ): String {
        val rawContent = loadRawChapterContent(chapter.id ?: "")
        // Junk `<p>` anti-scrape lồng lệch (truyenfull 260905) — strip trước khi
        // wrap `<p>` chuẩn, nếu không XHTML EPUB mismatch (sanitizer chỉ fix inline)
        // TXT: raw giữ HTML cho EPUB — phải đổi tag khối thành xuống dòng + strip tag
        // rác, không thì file text lẫn "<br>" literal suốt chương (dogfood 260906)
        val cleanedContent =
            if (format == ExportFormat.TEXT) {
                htmlToPlainText(cleanTextContent(rawContent))
            } else {
                stripParagraphTags(cleanTextContent(rawContent))
            }
        var formattedContent = formatChapterContent(chapter, cleanedContent, preserveParagraphs)
        if (format == ExportFormat.HTML) {
            formattedContent = sanitizeHtml(formattedContent)!!
        }
        return formattedContent
    }

    private fun loadRawChapterContent(chapterId: String): String = ((properties.savePath ?: "").toPath() / "raw/$chapterId$RAW_FILE_EXTENSION")
        .readTextOrNull(charset)
        ?: throw IOException("raw/$chapterId$RAW_FILE_EXTENSION")

    private fun cleanTextContent(content: String): String {
        var result = content
        for (trashRule: Trash in settings.trash) {
            if (trashRule.replace) {
                result = result.replace(unescapeTrashPattern(trashRule.src).toRegex(), unescapeTrashPattern(trashRule.to))
            }
        }
        return normalizeVietnamese(result)
    }

    /** HTML raw → text thuần cho TXT: `<br>`/block tag → xuống dòng, strip tag còn lại, decode entity. */
    private fun htmlToPlainText(content: String): String = stripParagraphTags(content)
        .replace(Regex("<br\\b[^>]*/?>", RegexOption.IGNORE_CASE), "\n")
        .replace(Regex("</?(div|p)\\b[^>]*>", RegexOption.IGNORE_CASE), "\n\n")
        .replace(Regex("<[^>]+>"), "")
        .replace("&nbsp;", " ")
        .replace("&amp;", "&")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", "\"")
        .replace(Regex("\\n{3,}"), "\n\n")

    private fun unescapeTrashPattern(pattern: String?): String = pattern?.replace("\\n", "\n")?.replace("\\r", "\r")?.replace("\\t", "\t") ?: ""

    private fun formatChapterContent(
        chapter: Chapter,
        content: String,
        preserveParagraphs: Boolean,
    ): String {
        var template = templateSyntax

        // Thay placeholder bằng giá trị thực
        template = replaceTemplatePlaceholder(template, properties.name, "\\[1\\]", "NAME")
        template = replaceTemplatePlaceholder(template, properties.author, "\\[2\\]", "AUTHOR")
        template = replaceTemplatePlaceholder(template, chapter.partName, "\\[3\\]", "NAME_PART")
        template = replaceTemplatePlaceholder(template, chapter.chapName, "\\[4\\]", "NAME_CHAP")
        template = template.replace("[ID]", chapter.id ?: "")

        // Format đoạn văn
        val formattedContent = formatParagraphs(content, preserveParagraphs)
        val paragraphPlaceholder = regexFind(template, "(\\[5\\].*\\[PARAGRAPH\\].*\\[5\\])", 1)
        return if (paragraphPlaceholder != null) {
            template.replace(paragraphPlaceholder, formattedContent)
        } else {
            template
        }
    }

    private fun replaceTemplatePlaceholder(
        template: String,
        value: String?,
        placeholderTag: String,
        placeholderKey: String,
    ): String {
        if (value == null || value.length < 2) {
            return template.replace("\\s*$placeholderTag.*?$placeholderTag".toRegex(), "")
        }
        val pattern = placeholderTag + "(.*?" + "\\[" + placeholderKey + "\\]" + ".*?)" + placeholderTag
        val replacement = template.replace(pattern.toRegex(), "$1")
        return replacement.replace("[" + placeholderKey + "]", value)
    }

    private fun formatParagraphs(
        contentIn: String,
        preserveParagraphs: Boolean,
    ): String {
        if (contentIn.isEmpty()) return contentIn
        var content = applyDropCapFormatting(contentIn)
        if (preserveParagraphs) {
            content = wrapInParagraphTags(content, extractParagraphTemplate())
        }
        return content
    }

    private fun extractParagraphTemplate(): String {
        val firstTag = regexFind(templateSyntax, "\\[5\\](.*)\\[PARAGRAPH\\](.*)\\[5\\]", 1) ?: ""
        val lastTag = regexFind(templateSyntax, "\\[5\\](.*)\\[PARAGRAPH\\](.*)\\[5\\]", 2) ?: ""
        return firstTag + "\n" + lastTag
    }

    private fun wrapInParagraphTags(
        content: String,
        paragraphTemplate: String,
    ): String {
        val parts = paragraphTemplate.split("\n")
        if (parts.size != 2) return content

        val firstTag = parts[0]
        val lastTag = parts[1]
        return firstTag + content.replace("\n", lastTag + "\n" + firstTag) + lastTag
    }

    private fun applyDropCapFormatting(content: String): String {
        if (content.isEmpty()) return content
        if (!settings.dropcapsEnabled || format != ExportFormat.HTML) return content

        val firstChar = content[0]
        if (firstChar == firstChar.lowercaseChar()) return content

        return settings.dropSyntax.replace("[DROP]", firstChar.toString()) + content.substring(1)
    }

    companion object {
        private const val RAW_FILE_EXTENSION = ".txt"
        private const val INTRODUCTION_ID = "gioithieu"
        private const val INTRODUCTION_NAME = "Giới Thiệu"
        private const val TABLE_OF_CONTENTS_TITLE = "Mục lục"
        private const val PROGRESS_MESSAGE_FORMAT = "[2/3]Exporting text..."
        private const val PROGRESS_MULTIPLIER = 100
    }
}
